frame/nfts/src/features/atomic_swap.rs

```diff
@@ -0,0 +1,152 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::*;
+use frame_support::{
+	pallet_prelude::*,
+	traits::{Currency, ExistenceRequirement::KeepAlive},
+};
+
+impl<T: Config<I>, I: 'static> Pallet<T, I> {
+	pub fn do_create_swap(
+		sender: T::AccountId,
+		collection_id: T::CollectionId,
+		item_id: T::ItemId,
+		desired_collection_id: T::CollectionId,
+		desired_item_id: T::ItemId,
+		maybe_price: Option<ItemPrice<T, I>>,
```

Comment:
I think the semantic of offering an NFT and *maybe* a price is weird.  
Now there can be no case in which i can buy an NFT just for a price.  
An NFT swap should also work without price, or is this common practice?

Reply:
> I think the semantic of offering an NFT _and_ a price is weird. Now there can be no case in which i can buy an NFT just for a price. An NFT swap should also work without price, or is this common practice?

The price is optional here, is it not?
It makes total sense to be able to add extra into the mix.
I'm curious though how this would work if the price needs to be added on the other end.
Looking at the logic here it seems that the price should have direction, as in who has to pay the added value - caller or the other party.

Reply:
The way this could be achieved is: 
`maybe_price`
`price_direction` -> enum (Caller, Receiver) (names are shit, but the idea is sound, because we would never need the price specified on both ends.)

a bit off-topic: BTW I've seen some decent ideas when interviewing with guys from Anoma a while ago. The extensions in their framework allow for sort of barter deals, what that basically means is:
we have 
A - wants some cabbage, sells some rice
B - wants some rice, sells some carrots
C - wants some carrots, sells some cabbage
If the quantities match - this transaction goes through and everybody gets and sells what they want. 

I know that this is not what we are doing here, but at the very least the idea of everybody using the same method call to say SELL/BUY is a good one, because it's basically a matter of direction only and then there should be a match. It's pretty much like crypto exchanges work too.

Reply:
I was thinking that to solve the order direction, a counterpart could simply register a swap :)

re your offtopic, the meta transactions could help here I think

Reply:
done